python/activator/middleware_interface.py

```diff
@@ -106,6 +107,39 @@ def get_central_butler(central_repo: str, instrument_class: str):
                   )
 
 
+def _transfer_multitype(dest_butler: Butler,
+                        source_butler: Butler,
+                        source_refs: collections.abc.Iterable[DatasetRef],
+                        *args, **kwargs) -> list[DatasetRef]:
+    """Efficiently transfer datasets of mixed types from one Butler
+    to another.
+
+    Parameters
+    ----------
+    dest_butler : `lsst.daf.butler.Butler`
+        The Butler to which the datasets are to be transferred.
+    source_butler : `lsst.daf.butler.Butler`
+        The Butler from which the datasets are to be transferred.
+    source_refs : iterable [`lsst.daf.butler.DatasetRef`]
+        Datasets to transfer from ``source_butler`` to ``dest_butler``.
+    *args, **kwargs
+        Additional arguments for `Butler.transfer_from`.
+
+    Returns
+    -------
+    transferred : `list` [`lsst.daf.butler.DatasetRef`]
+        The datasets transferred to ``dest_butler``.
+    """
+    refs_by_type = collections.defaultdict(set)
+    for ref in source_refs:
+        refs_by_type[ref.datasetType].add(ref)
+
+    transferred = []
+    for dtype, refs in refs_by_type.items():
```

Comment:
Is the idea that this means that the transaction period will be shorter because fewer files are transferred per call and so the overhead on the database locking will be reduced? 
It won't help with the file transfer speeds and will actively work against it if I enable lsst/daf_butler#1162?

Reply:
My understanding is that this prevents transactions from grabbing lots and lots of locks that they need, but don't need _at the same time_.

Reply:
It will definitely reduce the length and size of database transactions. It will, though, slow down the file transfers themselves once my ticket merges (since the point of my ticket is to do all the transfers in parallel [well with 10 workers]).

Reply:
I'm doing my tests on data.lsst.cloud on that ticket with 1 CPU.

Reply:
(and it turns out that 3 workers doing transfers is better than either 2 or 10 on Google). Suggesting the number of workers needs to be a parameter that can be tuned.